### ISSUE TITLE
JavaScript heap out of memory on v12.0.1

### ISSUE THREAD

**Issue by bausmeier** [guidesmiths/rascal#141]
Firstly, thanks for the package, it's been working well for us so far.

We're currently using v10.2.6 and experiencing an issue when trying to upgrade to v12.0.1. After upgrading the functionality still seems to be working as expected on our dev machines, but our tests have started to crash locally and on GitHub Actions with the following error:

```
<--- Last few GCs --->

[727712:0x558b398e0aa0]    72203 ms: Scavenge 2017.1 (2073.6) -> 2014.1 (2075.1) MB, 5.8 / 0.0 ms  (average mu = 0.831, current mu = 0.579) allocation failure
[727712:0x558b398e0aa0]    72295 ms: Scavenge 2018.6 (2075.1) -> 2016.3 (2078.4) MB, 6.3 / 0.0 ms  (average mu = 0.831, current mu = 0.579) allocation failure
[727712:0x558b398e0aa0]    72379 ms: Scavenge 2021.8 (2078.4) -> 2019.1 (2095.9) MB, 9.1 / 0.0 ms  (average mu = 0.831, current mu = 0.579) allocation failure


<--- JS stacktrace --->

FATAL ERROR: MarkCompactCollector: young object promotion failed Allocation failed - JavaScript heap out of memory
 1: 0x558b375212b1 node::Abort() [/usr/bin/node]
 2: 0x558b3743e144 node::FatalError(char const*, char const*) [/usr/bin/node]
 3: 0x558b376fb782 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
 4: 0x558b376fb9e8 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/bin/node]
 5: 0x558b378bb656  [/usr/bin/node]
 6: 0x558b378e7add v8::internal::EvacuateNewSpaceVisitor::Visit(v8::internal::HeapObject, int) [/usr/bin/node]
 7: 0x558b378f254e void v8::internal::LiveObjectVisitor::VisitBlackObjectsNoFail<v8::internal::EvacuateNewSpaceVisitor, v8::internal::MajorNonAtomicMarkingState>(v8::internal::MemoryChunk*, v
8::internal::MajorNonAtomicMarkingState*, v8::internal::EvacuateNewSpaceVisitor*, v8::internal::LiveObjectVisitor::IterationMode) [/usr/bin/node]
 8: 0x558b37912741 v8::internal::FullEvacuator::RawEvacuatePage(v8::internal::MemoryChunk*, long*) [/usr/bin/node]
 9: 0x558b378e2e25 v8::internal::Evacuator::EvacuatePage(v8::internal::MemoryChunk*) [/usr/bin/node]
10: 0x558b378e30f7 v8::internal::PageEvacuationTask::RunInParallel(v8::internal::ItemParallelJob::Task::Runner) [/usr/bin/node]
11: 0x558b378d4bc9 v8::internal::ItemParallelJob::Run() [/usr/bin/node]
12: 0x558b378f0169 void v8::internal::MarkCompactCollectorBase::CreateAndExecuteEvacuationTasks<v8::internal::FullEvacuator, v8::internal::MarkCompactCollector>(v8::internal::MarkCompactColle
ctor*, v8::internal::ItemParallelJob*, v8::internal::MigrationObserver*, long) [/usr/bin/node]
13: 0x558b37910c45 v8::internal::MarkCompactCollector::EvacuatePagesInParallel() [/usr/bin/node]
14: 0x558b37910f59 v8::internal::MarkCompactCollector::Evacuate() [/usr/bin/node]
15: 0x558b379118f3 v8::internal::MarkCompactCollector::CollectGarbage() [/usr/bin/node]
16: 0x558b378c8e9f v8::internal::Heap::MarkCompact() [/usr/bin/node]
17: 0x558b378c98a8 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/bin/node]
18: 0x558b378c9ea7 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/bin/node]
19: 0x558b378cd54c v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/bin/node]
20: 0x558b378cd5b4 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/bin/node]
21: 0x558b378924be v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [/usr/bin/node]
22: 0x558b3788c5a2 v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawArray(int, v8::internal::AllocationType) [/usr/bin/node]
23: 0x558b3788c654 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>, int, v8::internal::Handle<v8::internal::Oddball>, v8::int
ernal::AllocationType) [/usr/bin/node]
24: 0x558b37accfac v8::internal::OrderedHashTable<v8::internal::OrderedHashSet, 1>::Allocate(v8::internal::Isolate*, int, v8::internal::AllocationType) [/usr/bin/node]
25: 0x558b37acd17c v8::internal::OrderedHashTable<v8::internal::OrderedHashSet, 1>::Rehash(v8::internal::Isolate*, v8::internal::Handle<v8::internal::OrderedHashSet>, int) [/usr/bin/node]
26: 0x558b37acd7e1 v8::internal::OrderedHashTable<v8::internal::OrderedHashSet, 1>::Shrink(v8::internal::Isolate*, v8::internal::Handle<v8::internal::OrderedHashSet>) [/usr/bin/node]
27: 0x558b37bb8c66 v8::internal::Runtime_SetShrink(int, unsigned long*, v8::internal::Isolate*) [/usr/bin/node]
28: 0x558b37f33199  [/usr/bin/node]
```

This occurs after a test which usually takes the order of a few ms on v10.2.6 hangs for several seconds on v12.0.1. The memory usage of the app is usually low (~150MB) under normal operation, so I'm not sure what could be pushing it up so much. I don't want to just increase the heap size without understanding the underlying cause.

I was wondering if you may be able to give me some pointers as to where I should start before I dive deep into trying to debug this? We're using the namespace option during testing if that's relevant in any way.

**Comment by bausmeier** [comment]
Oh, and it's worth adding that the issue does *not* occur on v12.0.0 - it's specific to v12.0.1.

**Comment by cressie176** [comment]
Hi @bausmeier. Thanks for reporting. I'll review what changed in v12.0.1

**Comment by cressie176** [comment]
[this](https://github.com/guidesmiths/rascal/commit/dc34311bba941ff0e60225b9fe87322f4f3a15a4) is the only commit which had significant changes in v12.0.1. I can't see anything which would obviously cause a heap space issue. I don't suppose you're able to reproduce with in a public test case?

**Comment by bausmeier** [comment]
I'll see if I can extract some code which reproduces it to a public repo.

**Comment by bausmeier** [comment]
I started down the path of extracting a minimal reproduction from our codebase, but that turned out to be a lot of effort so I've abandoned that idea.

I decided to capture and compare heap snapshots instead and I've found something interesting, but haven't narrowed down the cause yet. Between the two snapshots I took *hundreds on thousands* of `TypeError`s with the message "Cannot read property 'on' of undefined" were added to the heap as `pendingUnhandledRejections`, and they seem to be coming from within rascal. Here's a screenshot of one of them, with one of the frames of the stack trace expanded:

![2021-02-23-150547_screenshot](https://user-images.githubusercontent.com/116090/108848618-dff2e780-75e9-11eb-90e3-c0824b546b0e.png)

I haven't pinpointed where it's coming from yet, but will continue to dig. It seems like something is trying to attach an event handler to an emitter which has already been cleaned up, and that the error might be triggering a tight or infinite loop.

**Comment by bausmeier** [comment]
It looks like it's likely related to the note in the README  under [Channel Pooling](https://github.com/guidesmiths/rascal#channel-pooling):

> Unfortunately there is a [bug](https://github.com/coopernurse/node-pool/issues/197#issuecomment-477862861) in generic-pool's implementation, which means that if the pool fails to create a channel, it can enter a tight loop, thrashing your CPU and potentially crashing your node process due to a memory leak. While we assess the long term use of pooling, we have put in a workaround. Errors will only be rejected after a configurable delay. This defaults to one second but can be overriden through the `rejectionDelayMillis` pool attribute.

And the comment linked to says:

> I think I might have found something that might explain this -- you need to make sure your `create` promise *does not fail*.
>
> I just spent some time debugging this issue, and it looks like `node-pool` will retry `create` in a tight loop, so if your pool does something expensive (which is why you're pooling in the first place), it will cause memory to run away.

So I suspect that the `TypeError` which I mentioned above is being thrown from within the `create` function, triggering that behaviour in the pool.

**Comment by bausmeier** [comment]
Is it possible that we need a check similar to:
https://github.com/guidesmiths/rascal/blob/5a1ff23f455ee747f96dd0d333d36c971def4dc3/lib/amqp/Subscription.js#L58-L61

(Which was [added](https://github.com/guidesmiths/rascal/commit/dc34311bba941ff0e60225b9fe87322f4f3a15a4#diff-c73aae904cb5c7df4cd735d952da7a41f88914cd81a77e319b949e949298f373R58-R61) in the commit you linked to above)

Before trying to attach the listeners to the channel when it is created:

https://github.com/guidesmiths/rascal/blob/5a1ff23f455ee747f96dd0d333d36c971def4dc3/lib/amqp/Vhost.js#L210-L211

I'm obviously not familiar with the code, but it looks like channel can be `undefined` at that point if this condition has been hit:
https://github.com/guidesmiths/rascal/blob/5a1ff23f455ee747f96dd0d333d36c971def4dc3/lib/amqp/Vhost.js#L326

Because my tests are creating and shutting down brokers quickly it's possible that some unrelated async code is trying to acquire a connection after the broker has already started shutting down. I will try and isolate the tests better so that doesn't happen, but it seems like this is an underlying bug.

**Comment by cressie176** [comment]
That's some excellent sleuthing. I strongly suspect you're right. I guess the options are to treat null as we do errors in the channel pool create code, or instead of yielding null, to yield an error and handle that elsewhere. I'll try to get to this asap.

**Comment by bausmeier** [comment]
Thanks, no rush getting to this - the previous version is working perfectly fine for us at the moment. We'd just like to try and keep our dependencies up to date.

I hacked some handling into the code to treat missing channels in the same way as errors in the `create` function, and that slowed the memory leak down, but didn't solve it. My tests ended up timing out instead of crashing, so at least that's some progress :laughing: There's potentially an issue with async handling in my tests as well, but I think that the `undefined` channel might need to be handled as a special case rather than an error.

Let me know if you need any more info or some help testing.

**Comment by cressie176** [comment]
I've pushed what I hope is a fix to GitHub, but not published to npm yet. The fix has two parts...

1. When the broker [forewarns](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L70-L76) a vhost of impending shutdown it calls [pauseChannelAllocation](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L371-L377), which pauses the channel creator queue and also the regular and confirm channel pools. After this it resumes the channel creator queue, which is necessary for reliable shutdown of subscribers

2. If the channel creator queue yields null, the channel pool will [defer a rejection](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L198-L201) as it does with errors. This should mean the pool doesn't enter a tight loop, although since the pools have been paused this may never happen.

I was only able to reproduce the error by shutting down the broker while publishing furiously. Even then I had to comment out a callback at the end of the [forewarn](https://github.com/guidesmiths/rascal/blob/master/lib/amqp/Vhost.js#L75) function before I saw the problem. Would you be willing to try the a rascal build from github and re-run your test please?

**Comment by bausmeier** [comment]
It works! :tada:

Pausing the channel allocation does the trick on its own, but deferring the rejection as well makes sense to me as an extra defensive mechanism.

Thanks so much for your help! I'm happy to close this issue with the next release, or now if you prefer.

**Comment by cressie176** [comment]
Published as rascal@12.0.3. Thank you very much for your help tracking this one down, it really was some excellent work on your part.
